### PR TITLE
#6313 - Make FaultModel:Fouling:Coil work

### LINKED ISSUE #5523: Operational Fault model for Coil Fouling appears to do nothing (v8.4)
## Issue overview

I've been experimenting with the coil fouling operational fault model in EnergyPlus v8.4, and have been having problems getting any difference in results from it.

Currently, I am trying to get the coil fouling module working with the 5ZoneAirCooled.idf example building that is included with EnergyPlus v8.4. I add the elements, select both of the cooling coils, and have put in what I consider appropriately small UA values (about 1/2 of the nominal UA value listed in the HTML file that is generated). [(Picture)](https://unmethours.com/upfiles/1455569140973535.png)

However, after running my "fouled" file, I get the exact same results as the original. I'd expect, at the least, to have increased pump energy from the reduced waterside dT across the coil.

Running the v8.3 5ZoneAirCooled.idf models (original and fouled) in v8.3 results in numbers I would expect. E.g. a slight increase in pump energy.

Reducing the UA values further in both cases results in the same qualitative results: v8.4 shows no difference, v8.3 shows a noticeable difference.

I know of no workarounds for users of the current version (8.4)

A verified fix would show differences in results between an example file and a file with a fouled cooling coil.
### Details

Some additional details for this issue (if relevant):
- Windows 7 64 bit
- EnergyPlus v8.4
### Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [ ] Defect file added (list location of defect file here)
- [ ] Ticket added to Pivotal for defect (development team task)
- [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### LINKED ISSUE #6313: Fault model coil fouling for hot water heating coil does not appear to be working
Issue overview
--------------
User file has a FaultModel:Fouling:Coil attached to a Coil:Heating:Water.  The coil output, water flow and all other results are identical with and without the fault model.  User reports that it also has no impact on cooling coils.

Another ticket complains about the cooling coil, with test file.

### Details
Some additional details for this issue (if relevant):
 - Platform Win64
 - Version of EnergyPlus v8.8.0 f4f8b9e466
 - Helpdesk ticket 12279 abd 12144

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added EnergyPlusDevSupport\DefectFiles
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #6313
 - Fixes #5523

`FaultModel:Fouling:Coil` Initial problem: `WaterCoils::GetWaterCoilInput()` tries to add an index 'FouledCoilID 'to `FaultsManager::FouledCoils` arrays here:

https://github.com/NREL/EnergyPlus/blob/8478cc45d5cb3988fb9b0025d33cc6a701afc21d/src/EnergyPlus/WaterCoils.cc#L944-L954

Problem is that when this happens, `FaultsManager::CheckAndReadFaults()` hasn't been called so the the `FouledCoils` array is always empty, so no fault are actually registered, ever.

Really, the Coil should be referencing the fault model too, to match other usages (Boiler, Chillers, EvapCoolers, etc) and so that you avoid having to constantly loop on all `FouledCoils` to find if there's one referenced your object.

I also implemented the calculation of the fouling factor in `FaultPropertiesFoulingCoil::CalFaultyCoilFoulingFactor()`, to centralized the fault code into the fault module, and updated kinks in the engineering reference guide (equation was wrong, probably just bad LaTeX, I found this wikipage on Devsupport [here](https://github.com/NREL/EnergyPlusDevSupport/wiki/NFPFaultModeling) that had the same formatting problems) and added a clarification in I/O ref/

I also improve a bit the error reporting when reading `FaultModel:Fouling:Coil`, and added specific Gtests for that. Added an extended Gtest to check coil/fault linking, fault input reading, and the calculation of the Fouling Factor as well.

**Note**: I realized there isn't a single file in `./testfiles` that includes `FaultModel:Fouling:Coil`. I'll ask on pivotal if I can create one.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by jmarrec** [comment]
Using the DefectFiles, the highlight being:

```
  Coil:Heating:Water,
    Air Loop Referenz AHU Heating Coil,  !- Name
    On 24/7,                 !- Availability Schedule Name
    6.64,                    !- U-Factor Times Area Value {W/K}
    [...]
    UFactorTimesAreaAndDesignWaterFlowRate,  !- Performance Input Method
    [...];

  Schedule:Compact,
    Fault Applicability Schedule,  !- Name
    Fraction,                !- Schedule Type Limits Name
    Through: 31 Dec,         !- Field 1
    For: Weekdays,           !- Field 2
    Until: 08:00,0,          !- Field 3
    Until: 18:00,1,          !- Field 4
    Until: 24:00,0,          !- Field 5
    For: Weekends SummerDesignDay WinterDesignDay AllOtherDays, !- Field 6
    Until: 24:00,0;          !- Field 7


  FaultModel:Fouling:Coil,
    FouledHeatingCoil,       !- Name
    Air Loop Referenz AHU Heating Coil,  !- Coil Name
    Fault Applicability Schedule,     !- Availability Schedule Name
    ,                        !- Severity Schedule Name
    FouledUARated,           !- Fouling Input Method
    3.32;                    !- UAFouled {W/K}

```

So Nominal UA is 6.64, Fouled UA is 3.32, and Fouling Occurs only on weekdays between 8:00 and 18:00. Before, you get absolutely no difference between the NoFault.idf and the Fault.idf. After you do.


image



Comparing with the "FaultSmallerUA" which uses a Fouling UA of 0.32 (instead of 3.32):

image

**Comment by nrel-bot-2** [comment]
@jmarrec it has been 8 days since this pull request was last updated.

**Comment by Myoldmopar** [comment]
@jmarrec this looks like a really great fix.  I pulled latest develop into this branch, cleaned up the unused variable warnings, and also brought in an unrelated initialization fix for the convection coefficient tests.  I'll give this a code review, but from the results you've shown, this is looking really great.

**Comment by Myoldmopar** [comment]
Outside of the unused variables warning, this looks good.  I'll take a look at the code, and if it is ready, I'll fix the unused variables and merge right after.

File: doc/engineering-reference/src/operational-faults.tex
Comment (Myoldmopar) [inline doc/engineering-reference/src/operational-faults.tex:121]: These fixes are very nice.

File: src/EnergyPlus/ConvectionCoefficients.cc
Comment (Myoldmopar) [inline src/EnergyPlus/ConvectionCoefficients.cc:4065]: This is a little fix I added to try to help the convection unit tests pass.  Unrelated to this work, move along...

File: src/EnergyPlus/FaultsManager.cc
Comment (Myoldmopar) [inline src/EnergyPlus/FaultsManager.cc:1804]: Interesting, and nice!

File: src/EnergyPlus/FaultsManager.cc
Comment (Myoldmopar) [inline src/EnergyPlus/FaultsManager.cc:1778]: I like that you are turning this around and more properly linking the fault with the coil.  If we proceed with adding more faults later, it would be good to give this a whole new design where the faults are classes with methods, and we can move away from this passive index linking.  But anyway, you've done a lot of good with this.  The faults should, well, work now!

File: src/EnergyPlus/FaultsManager.cc
Comment (Myoldmopar) [inline src/EnergyPlus/FaultsManager.cc]: Was this just accidentally left in?

File: src/EnergyPlus/OutputProcessor.hh
Comment (Myoldmopar) [inline src/EnergyPlus/OutputProcessor.hh:236]: @JasonGlazer can you approve this addition?

File: src/EnergyPlus/OutputProcessor.cc
Comment (Myoldmopar) [inline src/EnergyPlus/OutputProcessor.cc:5620]: @JasonGlazer it's converted from enum to string here.

File: src/EnergyPlus/OutputProcessor.cc
Comment (Myoldmopar) [inline src/EnergyPlus/OutputProcessor.cc:5773]: And from string to enum here.

File: src/EnergyPlus/WaterCoils.cc
Comment (Myoldmopar) [inline src/EnergyPlus/WaterCoils.cc:1985]: This is the actual calculations moved down here.  If another commit is necessary here, we should remove that commented function above.

File: src/EnergyPlus/FaultsManager.cc
Comment (Myoldmopar) [inline src/EnergyPlus/FaultsManager.cc:1805]: @mjwitte can you use your eagle eye on these output variable names?  Everything looks fine to me, but a second look would be good.

**Review by Myoldmopar — APPROVED** [review]
I'm OK with the bulk of the changes here.  I've asked @JasonGlazer to look at the units addition and @mjwitte to look at the output variable names.  I'm going to locally commit changes that eliminate the unused warning shown by CI, and also clean out the commented function I noted in my review.  If everything looks good and no other changes are made, I'll push my changes and merge without waiting on CI again.  If any functional changes are made I'll let CI have a final go at it.

File: src/EnergyPlus/OutputProcessor.cc
Comment (JasonGlazer) [reply src/EnergyPlus/OutputProcessor.cc:5620]: Good.

File: src/EnergyPlus/OutputProcessor.cc
Comment (JasonGlazer) [reply src/EnergyPlus/OutputProcessor.cc:5773]: Ok

File: src/EnergyPlus/OutputProcessor.hh
Comment (JasonGlazer) [reply src/EnergyPlus/OutputProcessor.hh:236]: Yes. Looks fine.

File: src/EnergyPlus/FaultsManager.cc
Comment (mjwitte) [reply src/EnergyPlus/FaultsManager.cc:1805]: @Myoldmopar These output variables are all good, the "U Factor Times Area Value" is consistent with the existing "Heating Coil U Factor Times Area Value" output and with the sizing report for other water coils.

File: src/EnergyPlus/FaultsManager.cc
Comment (JasonGlazer) [inline src/EnergyPlus/FaultsManager.cc:1796]: Good

**Review by JasonGlazer — APPROVED** [review]

File: src/EnergyPlus/WaterCoils.cc
Comment (mjwitte) [inline src/EnergyPlus/WaterCoils.cc:1955]: @jmarrec @JasonGlazer @Myoldmopar I'm a bit confused about what this `FaultyCoilFoulingFactor` means. I would have just reported a ratio of `UACoilVariable` (fouled) to `OriginalUACoilVariable`, but everyone else seems to be ok with this, so go ahead.

File: src/EnergyPlus/FaultsManager.cc
Comment (jmarrec) [reply src/EnergyPlus/FaultsManager.cc]: Yes. 
Originally I had made that its own routine and calculated here, which I liked better (Fault stuff contained in the Fault class itself)

But then I realized when checking and modifying the Cooling coil one that it definitely wasn't as simple and couldn't be well contained, so I resolved to move the calculation of the UAs (UAInternal / UAExternal) in the WaterCoils.cc  directly...

Comment (Myoldmopar) [reply src/EnergyPlus/FaultsManager.cc]: No problem.  I have my local commit ready with these commented lines deleted.

File: src/EnergyPlus/WaterCoils.cc
Comment (Myoldmopar) [reply src/EnergyPlus/WaterCoils.cc:1955]: @mjwitte thanks, I think we should for now proceed as-is and get this one behind us.

@jmarrec if anything needs to change here, open up an issue and continue addressing things there, for next release.  (Unless it's urgent, small, etc., etc., ...)

**Comment by Myoldmopar** [comment]
OK, thanks for the quick responses @mjwitte and @JasonGlazer .  I'm going to push my cleanup commit and merge this in.  Another one down.

**Pull request merged**
